File: 12-valid-parentheses/solution.ts
// Valid Parentheses
// Time: O(n), Space: O(n)

class Solution {
    isValid(s: string): boolean {
        const stack: string[] = [];
        const mapping: Record<string, string> = { ')': '(', '}': '{', ']': '[' };
        
        for (const char of s) {
            if (char in mapping) {
                const top = stack.length > 0 ? stack.pop()! : '#';
                if (mapping[char] !== top) {
                    return false;
                }
            } else {
                stack.push(char);
            }
        }
        
        return stack.length === 0;
    }
}

// Test
if (require.main === module) {
    const sol = new Solution();
    console.log(sol.isValid("()[]{}"));   // true
    console.log(sol.isValid("([)]"));     // false
    console.log(sol.isValid("{[]}"));     // true
}

export default Solution;

File: 12-valid-parentheses/solution.py
# Valid Parentheses
# Time: O(n), Space: O(n)

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', '}': '{', ']': '['}
        
        for char in s:
            if char in mapping:
                top = stack.pop() if stack else '#'
                if mapping[char] != top:
                    return False
            else:
                stack.append(char)
        
        return len(stack) == 0

# Test
if __name__ == "__main__":
    sol = Solution()
    print(sol.isValid("()[]{}"))   # True
    print(sol.isValid("([)]"))     # False
    print(sol.isValid("{[]}"))     # True
